opencog/cogserver/server/Agent.cc

```diff
@@ -88,13 +88,17 @@ std::string Agent::to_string() const
 void Agent::atomRemoved(AtomPtr atom)
 {
     Handle h = atom->getHandle();
-    for (size_t i = 0; i < _utilizedHandleSets.size(); i++)
-        _utilizedHandleSets[i].erase(h);
+    {
+        std::lock_guard<std::mutex> lock(_handleSetMutex);
+        for (size_t i = 0; i < _utilizedHandleSets.size(); i++)
+            _utilizedHandleSets[i].erase(h);
+    }
     removeAtomStimulus(h);
```

Comment:
Although this change is correct, the idea of using notifier chains for this purpose is terrible: it has a large performance impact, and is not fundamentally scalable.   The very concept on which this is built needs to be rethought and redesigned.